### PR TITLE
[WIP] Forced hosts configuration support

### PR DESCRIPTION
**Pull request by mikroskeem**
Solves #116 

TODO:
- [x] Configuration bits
- [x] More fool-proof configuration reading
- [x] Hook forced hosts up to Velocity internals
- [x] Should invalid server entries get removed internally?

### CODE REVIEW

File: proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java [hunk 1]
```diff
@@ -118,6 +118,14 @@ public void start() {
             servers.register(new ServerInfo(entry.getKey(), AddressUtil.parseAddress(entry.getValue())));
         }
 
+        for (Map.Entry<String, List<String>> entry : configuration.getForcedHosts().entrySet()) {
+            for (String server : entry.getValue()) {
+                if (!servers.getServer(server).isPresent()) {
+                    logger.warn("Server '{}' for forced host '{}' does not exist", server, entry.getKey());
```
Comment (mikroskeem) [inline proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java]: IIRC invalid server entries cause premature disconnection from proxy, e.g if 3rd server is invalid and there are 7 servers in list, then player would get disconnected on 3rd server connection attempt.

So should list be copied and rewritten here or should logic be changed where disconnection happens?

Comment (astei) [reply proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java]: Your concern is unfounded. Velocity will try to connect to all seven servers, and only after all attempts are exhausted will it disconnect the server.

I would prefer this were a fatal error that would prevent the configuration from being loaded.

File: proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java [hunk 2]
```diff
@@ -118,6 +118,16 @@ public void start() {
             servers.register(new ServerInfo(entry.getKey(), AddressUtil.parseAddress(entry.getValue())));
         }
 
+        for (Map.Entry<String, List<String>> entry : configuration.getForcedHosts().entrySet()) {
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java]: This check should be performed in `VelocityConfiguration.validate()`.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java [hunk 3]
```diff
@@ -344,7 +347,11 @@ private void handleConnectionException(RegisteredServer rs, @Nullable Component
     }
 
     Optional<RegisteredServer> getNextServerToTry() {
-        List<String> serversToTry = server.getConfiguration().getAttemptConnectionOrder();
+        if (serversToTry == null) {
+            Optional<List<String>> forcedHosts = getVirtualHost().map(vHost -> server.getConfiguration().getForcedHosts().get(vHost.getHostString()));
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java]: This code will trigger the Optional Checker in Checker Framework. Prefer a `virtualHost != null` check and `server.getConfiguration().getForcedHosts().getOrDefault(virtualHost, server.getConfiguration().getAttemptConnectionOrder())`.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java [hunk 4]
```diff
@@ -62,6 +64,7 @@
     private final InetSocketAddress virtualHost;
     private GameProfile profile;
     private PermissionFunction permissionFunction = null;
+    private List<String> serversToTry = null;
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java]: I forgot! This should be marked `@MonotonicNonNull`.

File: proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java [hunk 5]
```diff
@@ -421,6 +443,61 @@ public String toString() {
...
+        private ForcedHosts() {}
+
+        private ForcedHosts(Toml toml) {
+            if (toml != null) {
+                Map<String, List<String>> forcedHosts = new HashMap<>();
+                for (Map.Entry<String, Object> entry : toml.entrySet()) {
+                    if (entry.getValue() instanceof String) {
+                        forcedHosts.put(stripQuotes(entry.getKey()), ImmutableList.of((String) entry.getValue()));
+                    } else if (entry.getValue() instanceof List) {
+                        forcedHosts.put(stripQuotes(entry.getKey()), ImmutableList.copyOf((List<String>) entry.getValue()));
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java]: Make sure the list is not empty.

**Pull request merged**